### PR TITLE
Prometheus: Fix some of the alerting queries that use reduce/math operation

### PR DESCRIPTION
**Pull request by ivanahuckova**
**What this PR does / why we need it**:
In https://github.com/grafana/grafana/pull/43622 we have fixed filling out of missing values with null. However, this affected alerting and customers as they build upon the old behaviour, were missing datapoints in prometheus would not be replaced by null. The new behaviour is a bit problematic for now, as any reduce/math operation will return NaN if any null value was passed as part of the input, resulting in triggering an alert. There is a PR that is supposed to introduce Mode to drop [NaN/Inf/Null in Reduction operations]( https://github.com/grafana/grafana/pull/43583) that would fix this issue if users have set that mode, but for now we want to make sure that we are not breaking alerting and we are adding old-way of processing matrix results without filling out nulls for alerting. 

How to test: 
1. Run tns-demo prometheus
2. Create alerting query such as `go_gc_duration_seconds{job="monitoring/node_exporter"}` and step 5s
3. Turn off prom/node-exporter docker image for minute
4. Turn it on
5. Run a query, there shouldn't be a gap and all points should be connected in graph

### CODE REVIEW

File: pkg/tsdb/prometheus/time_series_query.go
Comment (JohnnyQQQQ) [inline pkg/tsdb/prometheus/time_series_query.go:205]: WDYT about renaming `fromAlerting` to something _like_ `fillNulls` and then using a pair of constants to set it? This would result in much more readable code.

Comment (leeoniya) [reply pkg/tsdb/prometheus/time_series_query.go:205]: i was about to write the same comment.

`FromAlert` is too specific, while something like `FillNulls` can be reused and more straightforward.

**Review by JohnnyQQQQ — APPROVED** [review]
I tested it locally and it works fine, so from the alerting side everything is perfect 👍  I left a nit regarding readability.

File: pkg/tsdb/prometheus/time_series_query.go
Comment (ivanahuckova) [reply pkg/tsdb/prometheus/time_series_query.go:205]: Great call! I'll update it.

File: pkg/tsdb/prometheus/time_series_query.go
Comment (toddtreece) [inline pkg/tsdb/prometheus/time_series_query.go:132]: should this be handled on the frontend when the request is for a time series panel, or maybe by a new `Format` option in the query?

it seems like this might also be needed for the `Table` format. here's an example from this branch:
image

vs `8.2.4`:
image

**Comment by ivanahuckova** [comment]
> should this be handled on the frontend when the request is for a time series panel, or maybe by a new Format option in the query?

@toddtreece Yes, I kind of agree at this point. When we initially decided to implement it on backend, it was because we wanted to receive same data frames from any query (alerting/expressions/frontend). However, as this is not currently working for alerting and the null filling shouldn't be for table results (because as you pointed out, it produces different view then in 8.2.4), I think we should remove null adding from backend and use [something similar to this](https://github.com/grafana/grafana/blob/main/public/app/plugins/datasource/prometheus/result_transformer.ts#L387) [here in transformV2](https://github.com/grafana/grafana/blob/main/public/app/plugins/datasource/prometheus/result_transformer.ts#L102)

**Comment by leeoniya** [comment]
i agree that any `null`-filling should happen on the frontend, since it will be viz-dependent.

what we need is a generic transformer that does the opposite of "Connect null values < Threshold" transform, which would be something like "Disconnect values > Threshold". maybe we can even roll it into the same UI somehow.

image

this has been kind of requested before, and is useful outside of the prometheus/missing-steps case. see https://github.com/grafana/grafana/issues/34542#issuecomment-886069580.

once we have this, we can apply it auto-magically to prometheus responses in the timeseries viz, but we will need to know the expected `step` size to set as the threshold input to this transformer.

we propose to pass this in the field's metadata: https://github.com/grafana/grafana-plugin-sdk-go/pull/453

**Comment by ivanahuckova** [comment]
I am closing this PR as we agreed with @toddtreece that the null filling should be removed from backend and moved to frontend and that he will take this over.

**Comment by toddtreece** [comment]
@aocenas we are working on a frontend fix for this, but to unblock alerting, i was thinking of merging this as a temp fix and reverting once the frontend change is ready to merge. does that sound ok?

**Pull request merged**
